### ISSUE TITLE
Save ~360,000 gas on collection creation by packing 2 pixels into one uint8

### ISSUE THREAD

**Issue by assafom** [Steemhunt/dixel-v2-contract#16]
The smallest uint type in solidity is uint8, and at the moment this is what Dixel2 uses to save each pixel's value.
This gives us 256 values to save in each uint8; but we need only 16, as the pallete size is 16.

This means that by using bit operations, we can utilize each uint8 to save 2 pixels.
For example, pixels[0] would be the colors of both the first and second pixel. The rightmost 4 bits would be the first pixel, and the leftmost 4 bits would be the second pixel.

While this will add some complexity to the front end and back end in creating/reading the pixels array, this saves 360,000 gas upon collection creation - a significant amount. (It is up to you to decide whether it is worth it.)

Basically, the formulas for getting the pixel values at `_generateSVG` (the only place that reads the values) will be:
```
uint256 prev = pixels[y * CANVAS_SIZE / 2] & 15;
uint256 current = (pixels[y * CANVAS_SIZE/2 + x/2] >> 4*(x%2)) & 15;
```
In a few hours I will try to create a patch/PR.

**Comment by assafom** [comment]
Note: as now the each pixel is uint8, the pixels array accepts values that are bigger than 15.
This means that if the user for example submits a pixel with color value 20, it will succeed in creating the contract, but then *the contract will revert when trying to create the metadata or SVG*, because there will be an out of bounds exception, as the contract will try to read pallete[20], but pallete only goes up to 15.
So the user would have spent funds for deployment, but the contract will not work.
I am not sure if it worthy to mitigate this by checking the input, as this is unlikely to happen.
But thought it's worth bringing it to your attention.
If we use the packed array method I detailed above, this problem will also be fixed, because then the user couldn't submit a pixel with a color larger than 15.

**Comment by assafom** [comment]
Implemented in PR #17.

**Comment by sebayaki** [comment]
This is brilliant! 
Thanks a lot for your effort on looking at the code deeply. 🙏🏻

Your PR is merged and our front-end dev is working on changes on his end.
